ykrt/src/compile/j2/opt/known_bits.rs

```diff
@@ -0,0 +1,482 @@
+    /// E.g. `unknown & (~1)` zeroes the least significant bit.
+    /// We thus define information gain as the knowledge of a set zero.
+    fn opt_and(&mut self, opt: &mut PassOpt, mut inst: And) -> OptOutcome {
+        inst.canonicalise(opt);
+        let And { tyidx, lhs, rhs } = inst;
+        let lhs_b = self.as_knownbits(opt, lhs);
+        let rhs_b = self.as_knownbits(opt, rhs);
+        let res = lhs_b.and(&rhs_b);
+        self.set_knownbits(res.clone());
+
+        // If we know the output's bits, emit that.
+        if res.all_known() {
+            return OptOutcome::Rewritten(Inst::Const(Const {
+                tyidx,
+                kind: ConstKind::Int(res.as_int()),
+            }));
+        }
+
+        // If no new information (set zeroes) was gained, that means this
+        // op is useless.
+        if rhs_b.all_known()
+            && (rhs_b
+                .zeroes()
+                .bitand(&lhs_b.known_ones().bitor(&lhs_b.unknowns))
+                .count_ones()
+                == 0)
+        {
+            return OptOutcome::Equiv(lhs);
+        }
+
+        OptOutcome::Rewritten(inst.into())
+    }
+
+    fn opt_const(&mut self, inst: Const) -> OptOutcome {
+        let Const { tyidx: _, kind } = &inst;
+        if let ConstKind::Int(kind) = kind {
+            self.set_knownbits(KnownBitValue::from_constant(kind))
+        }
+        OptOutcome::Rewritten(inst.into())
+    }
+
+    /// the `or` operation adds new information in the form of set ones.
+    /// E.g. `unknown | 01` selects the least significant bit.
+    /// We thus define information gain as the knowledge of a set one.
+    fn opt_or(&mut self, opt: &PassOpt, mut inst: Or) -> OptOutcome {
+        inst.canonicalise(opt);
+        let Or {
+            tyidx,
+            lhs,
+            rhs,
+            disjoint: _,
+        } = inst;
+        let lhs_b = self.as_knownbits(opt, lhs);
+        let rhs_b = self.as_knownbits(opt, rhs);
+        let res = lhs_b.or(&rhs_b);
+        self.set_knownbits(res.clone());
+
+        // If we know the output's bits, emit that.
+        if res.all_known() {
+            return OptOutcome::Rewritten(Inst::Const(Const {
+                tyidx,
+                kind: ConstKind::Int(res.as_int()),
+            }));
+        }
+
+        // If no new information (set ones) was gained, that means this
+        // op is useless.
+        if rhs_b.all_known()
+            && (rhs_b
+                .known_ones()
+                .bitand(&lhs_b.zeroes().bitor(&lhs_b.unknowns))
+                .count_ones()
+                == 0)
+        {
+            return OptOutcome::Equiv(lhs);
+        }
+
+        OptOutcome::Rewritten(inst.into())
+    }
+}
+
+/// Known bits for a single value.
+///
+/// This uses the same bit representation found in
+/// https://pypy.org/posts/2024/08/toy-knownbits.html#the-knownbits-abstract-domain
+///
+/// In short:
+/// one     unknown     knownbit
```

Comment:
Probably best to either use Markdown formatting for the table (or, at a push, a code block).

Reply:
I would also probably make the first row the `?` state since that's the first. Then it becomes clear that one can go from row 1 to either rows 2 or 3, but no other transitions are legal (so we can condense the slightly long lattice comment perhaps?).